### PR TITLE
Restoring backward compatibility, refactoring

### LINKED ISSUE #83: Retry decorator consumes so much time when we have large number of requests from unit
It seems to the retry decorator **consumes so much time when we have large number of requests from unit**, it is working fine if we have a small number requests.

Here some statistics i observed using robot profiler

With retry decorator statistics 

<pre>
1    0.000    0.000    0.000    0.000 RequestsKeywords.py:21(__init__)
  248    0.011    0.000  287.346    1.159 RequestsKeywords.py:217(patch_request)
102705    0.816    0.000    4.060    0.000 RequestsKeywords.py:25(_utf8_urlencode)
    4    0.000    0.000    0.002    0.000 RequestsKeywords.py:37(_create_session)
  469    0.008    0.000  431.145    0.919 RequestsKeywords.py:387(_get_request)
<b>101988    0.002    0.000  84256.164    0.809 RequestsKeywords.py:34(f_retry)</b>
101988    2.587    0.000 84456.164    0.828 RequestsKeywords.py:399(_post_request)
  248    0.009    0.000  287.253    1.158 RequestsKeywords.py:412(_patch_request)
</pre>

Instead I used ```requests``` retry [method](http://www.coglib.com/~icordasc/blog/2014/12/retries-in-requests.html), it was much faster than retry decorator

``` Python
    if max_retries > 0:
            a = requests.adapters.HTTPAdapter(max_retries=max_retries)
            s.mount('https://', a)
            s.mount('http://', a)
```
without retry decorator statistics 

``` Shell

        1    0.000    0.000    0.000    0.000 RequestsKeywords.py:21(__init__)
      248    0.011    0.000  287.346    1.159 RequestsKeywords.py:217(patch_request)
   102705    0.816    0.000    4.060    0.000 RequestsKeywords.py:25(_utf8_urlencode)
        4    0.000    0.000    0.002    0.000 RequestsKeywords.py:37(_create_session)
      469    0.008    0.000  431.145    0.919 RequestsKeywords.py:387(_get_request)
   101988    2.587    0.000 84456.164    0.828 RequestsKeywords.py:399(_post_request)
      248    0.009    0.000  287.253    1.158 RequestsKeywords.py:412(_patch_request)
```

@bulkan could you please comment on this?

### LINKED ISSUE #84: Post Request and Put Request does not support parameter passsing
I was trying to issue `POST / PUT` request to `/uri?key=value` with body in `xml` format but it seems that both `POST` and `PUT` request does not support passing of parameters `key=value`.

I have tried the following:
```
${resp}=    Post Request    session1    /uri    params=key=value    data=${body}
```
The following error is encountered:
`TypeError: post_request() got an unexpected keyword argument 'params'`

Is there other way to pass parameters in `POST/PUT` request?

### PR DESCRIPTION
**Pull request by oleduc**
This pull requests is a response to perceived source decay. Its mostly de-duplication and cleanup but I also had to revert a few things @vkosuri did.

Here is a breakdown of what I've done and why:

**1 - Removed retry decorator**
I remove it because of two reasons. Firstly because it was not very fast and it was implemented using a hard sleep. Also requests already supports adapter level max_retries. Secondly, the decorator would break Robotframework's parameter translation, which in turn made it impossible to use strings containing the equal symbol as it would be interpreted as a keyword argument.

**2- Removed overly broad catch around requests calls**
It would cause every exception to be interpreted as a retry exception. Also its not necessary, we want requests exceptions to reach the framework if possible.

**3- Deduplication**
A lost of the debug code was just copy pasted and needed to be refactored.

**4- Made arguments optional again**
Because it broke backward compatibility.

Also @bulkan if you want help with code reviews and maintenance, me and my employers would gladly help around if you provided us with collaborator access :)

Closes #84 
Closes #83
Includes #85

### CODE REVIEW

**Comment by vkosuri** [comment]
Few more comments

If you are not using requests exception, it doesn't required

``` LINE 12: from requests.exceptions import RequestException```

timeout should in float, because robot data is in unicode format (not always)

``` LINE 78: self.timeout = float(timeout) ```

LINE 175 and 177 does it require?

```  `max_delay` The maximum number of delay each connection should attempt.
        
        `max_backoff` Expoential backoff
```

Does debug messages are printing into robot log?

**Comment by oleduc** [comment]
1 - For the max_retry, I'm not sure what is the use case for more granular control over max_retry for the same session.
2 - I remove unused impors.
3 - Good point, I corrected that.
**LINE 175 and 177 does it require?**
No I removed it.
**Does debug messages are printing into robot log?**
Yes, I didn't change the mechanism you implemented, I just extracted it into a function.

**Comment by oleduc** [comment]
@vkosuri Does this look good to you?

**Comment by vkosuri** [comment]
@oleduc thank you nice work, 

@bulkan any comments from your side?

@oleduc if possible could you also add some testcases for url params for post, patch and delete after this merge? and documentation generation?

**Comment by oleduc** [comment]
@vkosuri Yes, good idea.

**Comment by vkosuri** [comment]
@oleduc working on this issue https://github.com/bulkan/robotframework-requests/issues/83

**Comment by vkosuri** [comment]
@oleduc few more comments on this PR

  Instead print messages for deprecation warning i think it is beter to having logger.warn

``` Python 
print "Deprication Warning  Use Get Request in the future"
```
``` Python 
logger.warn("Deprication Warning  Use Get Request in the future")
```

  Enable reStructed for doc strings for all deprecated doc string

```* * *   Deprecated- See Post Request now   * * *```

```**Deprecated- See Post Request now**```

  We are using last_resp for history, i think this is not required

```(@Kosuri Why?)```

  doc string position changes, even though we are using named arguments, i think it is better to having in proper position

``` LINE 349: `params` url parameters to append to the uri````


  def post,  patch and delete doesn't having params, could you please add params arguments these procedures also?

``` Python
def post(self, alias, uri, data={}, headers=None, files=None, allow_redirects=None, timeout=None):
 def patch(self, alias, uri, data={}, headers=None, files={}, allow_redirects=None, timeout=None):
def put(self, alias, uri, data=None, headers=None, allow_redirects=None, timeout=None):
def delete(self, alias, uri, data=(), headers=None, allow_redirects=None, timeout=None):
```

**Comment by vkosuri** [comment]
https://github.com/bulkan/robotframework-requests/pull/81

**Comment by vkosuri** [comment]
@oleduc any updates on this?

@bulkan any comments on this PR?

**Comment by bulkan** [comment]
I just came back from holidays will look at this in a day or two

**Comment by vkosuri** [comment]
@bulkan any comments on this PR?

**Comment by bulkan** [comment]
@vkosuri once @oleduc updates PR with your last suggestion it looks good to merge

**Comment by oleduc** [comment]
@vkosuri ***def post, patch and delete doesn't having params, could you please add params arguments these procedures also?***

These methods are deprecated, we should keep their API as is.

**Comment by vkosuri** [comment]
@oleduc thank you agreed with you

@bulkan it looks good to me, could merge this PR?

File: src/RequestsLibrary/RequestsKeywords.py [hunk 1]
```diff
@@ -325,20 +253,18 @@ def get(self, alias, uri, headers=None, params={}, allow_redirects=None, timeout
         `headers` a dictionary of headers to use with the request
         
         `timeout` connection timeout
-        """        
+        """
         print "Deprication Warning  Use Get Request in the future"
```
Comment (bulkan) [inline src/RequestsLibrary/RequestsKeywords.py]: this is still here, can we change to logger.warn ?

**Comment by oleduc** [comment]
@bulkan Sorry I didn't have time to do it today. I'll try to have it done by 03/02/2016 @ 23:00 GMT-5

**Comment by vkosuri** [comment]
@oleduc any updates?

**Comment by bulkan** [comment]
@vkosuri if i merge this can you fix up the `print` to `warn` ?

**Comment by vkosuri** [comment]
@bulkan yeah sure

**Comment by bulkan** [comment]
@vkosuri ok merged :)

**Comment by oleduc** [comment]
@bulkan Sorry, I can do it really quickly.

**Pull request merged**